fmt/src/selection.rs

```diff
@@ -182,15 +180,45 @@ fn select_files_with_git(
         builder.build().context(SelectWithIgnoreSnafu)?
     };
 
-    let mut it = WalkDir::new(basedir).into_iter();
+    let basedir = basedir
+        .canonicalize()
+        .with_context(|_| ResolveAbsolutePathSnafu {
+            path: basedir.display().to_string(),
+        })?;
+    let mut it = WalkDir::new(basedir).follow_links(false).into_iter();
+
+    let workdir = repo
+        .work_dir()
+        .context(GixPathNotFountSnafu { path: "workdir" })?;
+    let workdir = workdir
+        .canonicalize()
+        .with_context(|_| ResolveAbsolutePathSnafu {
+            path: workdir.display().to_string(),
+        })?;
+    let worktree = repo
+        .worktree()
+        .context(GixPathNotFountSnafu { path: "worktree" })?;
+    let mut excludes = worktree
+        .excludes(None)
+        .map_err(Box::new)
+        .context(GixExcludeOpSnafu)?;
+
     while let Some(entry) = it.next() {
         let entry = entry.context(TraverseDirSnafu)?;
         let path = entry.path();
         let file_type = entry.file_type();
+
+        let rela_path = path
+            .strip_prefix(&workdir)
+            .expect("git repository encloses iteration");
+        let platform = excludes
+            .at_path(rela_path, Some(file_type.is_dir()))
```

Comment:
This is why this is fast - after all the initialization, only the necessary work is performed here. All the pre-amble is hidden in `git2`, but that does only mean it does a lot of work for each call. Think about the relative-path requirement, it's the same for `git2` and it will canonicalize everything because it has to, which is really expensive as well.

That's also the reason `gix` won't offer such a nice and convenient API like `git2`, it's simply too wasteful (or very stateful internally, which comes with its own problems).

Reply:
Agree.

Out of the thread, in another project I participate we discover an issue that PROST's performance is even "worse" the go-protobuf, cause it makes less assumption/requirement on ownership/lifetime so it has to copy the buffer internally.

* https://github.com/GreptimeTeam/greptimedb/pull/3478
* https://github.com/GreptimeTeam/greptimedb/pull/3425